src/librustc_codegen_llvm/back/lto.rs

```diff
@@ -111,29 +110,47 @@ fn prepare_lto(
             }
 
             let archive = ArchiveRO::open(&path).expect("wanted an rlib");
-            let bytecodes = archive
+            let obj_files = archive
                 .iter()
                 .filter_map(|child| child.ok().and_then(|c| c.name().map(|name| (name, c))))
-                .filter(|&(name, _)| name.ends_with(RLIB_BYTECODE_EXTENSION));
-            for (name, data) in bytecodes {
+                .filter(|&(name, _)| looks_like_rust_object_file(name));
+            for (name, data) in obj_files {
                 let _timer =
                     cgcx.prof.generic_activity_with_arg("LLVM_lto_load_upstream_bitcode", name);
-                info!("adding bytecode {}", name);
-                let bc_encoded = data.data();
-
-                let (bc, id) = match DecodedBytecode::new(bc_encoded) {
-                    Ok(b) => Ok((b.bytecode(), b.identifier().to_string())),
-                    Err(e) => Err(diag_handler.fatal(&e)),
-                }?;
-                let bc = SerializedModule::FromRlib(bc);
-                upstream_modules.push((bc, CString::new(id).unwrap()));
+                info!("adding bitcode from {}", name);
+                match get_bitcode_slice_from_object_data(data.data()) {
+                    Ok(bc) => {
+                        let bc = SerializedModule::FromRlib(bc.to_vec());
+                        upstream_modules.push((bc, CString::new(name).unwrap()));
+                    }
+                    Err(e) => return Err(diag_handler.fatal(&e)),
+                }
             }
         }
     }
 
     Ok((symbol_white_list, upstream_modules))
 }
 
+fn get_bitcode_slice_from_object_data(obj: &[u8]) -> Result<&[u8], &'static str> {
+    let mut len = 0;
+    let data =
+        unsafe { llvm::LLVMRustGetBitcodeSliceFromObjectData(obj.as_ptr(), obj.len(), &mut len) };
+    if !data.is_null() {
+        assert!(len != 0);
+        let bc = unsafe { slice::from_raw_parts(data, len) };
+
+        // `bc` must be a sub-slice of `obj`.
+        assert!(obj.as_ptr() <= bc.as_ptr());
+        assert!(bc[bc.len()..bc.len()].as_ptr() < obj[obj.len()..obj.len()].as_ptr());
+
+        Ok(bc)
+    } else {
+        assert!(len == 0);
+        Err("failed to extract bitcode from object file for LTO")
```

Comment:
Could this use `last_error()` to pull out the LLVM error message as well?

Reply:
Yes, done.